homeworks/max.zinchenko_zzeenj/homework_6/task_2.js

```diff
@@ -0,0 +1,81 @@
+class Horse {
+    constructor(name, breed) {
+        this.name = name;
+        this.breed = breed;
+    }
+}
+class Racer extends Horse {
+    constructor(name, breed) {
+        super(name, breed);
+        this.distance = 0;
+        this.speed = 0;
+        this.setSpeed();
+    }
+
+    setSpeed() {
+        this.speed = Math.floor(Math.random() * (15 - 10 + 1) + 10);
+    }
+
+    run() {
+        for (let i = 1; i <= 10; i++) {
+            setTimeout(() => {
+                this.distance += this.speed;
+                this.setSpeed();
+            }, 1000 * i);
+        }
+    }
+}
+
+class Race {
+    constructor() {
+        this.horses = [];
+    }
+
+    createRace() {
+        const allHorses = [{ name: 'Agape', breed: 'Jutland horse' },
+            { name: 'Gina', breed: 'Lithuanian Heavy Draught' },
+            { name: 'Darcy', breed: 'Novokirghiz' },
+            { name: 'Wriggley', breed: 'Morab' },
+            { name: 'Nelly', breed: 'Russian Don' },
+            { name: 'Seyton', breed: 'San Fratello' },
+            { name: 'Nor', breed: 'Sorraia' },
+            { name: 'Gloryl', breed: 'Quarab' },
+            { name: 'Aida', breed: 'Messara' },
+            { name: 'Deputy', breed: 'Warlander' }];
+        allHorses.forEach((racer) => {
+            this.horses.push(new Racer(racer.name, racer.breed));
+        });
+    }
+
+    startRace() {
+        setTimeout(() => {
+            setTimeout(() => {
+                setTimeout(() => {
+                    setTimeout(() => {
+                        this.horses.forEach(racer => racer.run());
+                        for (let i = 1; i <= 5; i++) {
+                            setTimeout(() => {
+                                this.horses.forEach(racer => console.log(`Name: ${racer.name} Breed: ${racer.breed} Distance: ${racer.distance}`));
+                            }, 2000 * i);
+                        }
+                    }, 0);
+                    console.log('Go!');
+                }, 400);
+                console.log('Set');
+            }, 300);
+            console.log('Ready');
+        }, 200);
```

Comment:
although, idea is cool, but I personally would never use this pattern...

you could create a function returning a promise, which could make this code look much more readable

e.g.,

`
function fireEvent(event, timeout) {
    return new Promise(resolve => {
        setTimeout(() => {
            console.log(event);
            resolve(event);
        }, timeout)
    })
}
`

then you could just chain promises in a single line or event better `await` for promise chain

the Christmas tree is still there...